packages/orchestrator/lib/clients/processor.ts

```diff
@@ -0,0 +1,145 @@
+
+const logger = getLogger('orchestrator.clients.processor');
+
+export class OrchestratorProcessor {
+    private handler: (task: OrchestratorTask) => Promise<Result<JsonValue>>;
+    private groupKey: string;
+    private orchestratorClient: OrchestratorClient;
+    private queue: Queue;
+    private stopped: boolean;
+    private abortControllers: Map<string, AbortController>;
+    private terminatedTimer: NodeJS.Timeout | null = null;
+    private checkForTerminatedInterval: number;
+
+    constructor({
+        handler,
+        opts
+    }: {
+        handler: (task: OrchestratorTask) => Promise<Result<JsonValue>>;
+        opts: { orchestratorClient: OrchestratorClient; groupKey: string; maxConcurrency: number; checkForTerminatedInterval?: number };
+    }) {
+        this.stopped = true;
+        this.handler = handler;
+        this.groupKey = opts.groupKey;
+        this.orchestratorClient = opts.orchestratorClient;
+        this.queue = new Queue(opts.maxConcurrency);
+        this.abortControllers = new Map();
+        this.checkForTerminatedInterval = opts.checkForTerminatedInterval || 1000;
+    }
+
+    public start() {
+        this.stopped = false;
+        this.terminatedTimer = setInterval(async () => {
+            await this.checkForTerminatedTasks();
+        }, this.checkForTerminatedInterval); // checking for cancelled/expired doesn't require to be very responsive so we can do it on an interval
+        void this.processingLoop();
+    }
+
+    public stop() {
+        this.stopped = true;
+        if (this.terminatedTimer) {
+            clearInterval(this.terminatedTimer);
+        }
+    }
+
+    private async checkForTerminatedTasks() {
+        if (this.stopped || this.abortControllers.size <= 0) {
+            return;
+        }
+        const ids = Array.from(this.abortControllers.keys());
+        const search = await this.orchestratorClient.search({ ids });
+        if (search.isErr()) {
+            return Err(search.error);
+        }
+        for (const task of search.value) {
+            // if task is already in a terminal state, invoke the abort signal
+            if (['FAILED', 'EXPIRED', 'CANCELLED', 'SUCCEEDED'].includes(task.state)) {
+                const abortController = this.abortControllers.get(task.id);
+                if (abortController) {
+                    abortController.abort();
+                    this.abortControllers.delete(task.id);
+                }
+            }
+        }
+        return;
+    }
+
+    private async processingLoop() {
+        while (!this.stopped) {
+            if (this.queue.available() > 0) {
+                const tasks = await this.orchestratorClient.dequeue({ groupKey: this.groupKey, limit: this.queue.available() * 2, waitForCompletion: true }); // fetch more than available to keep the queue full
+                if (tasks.isErr()) {
+                    logger.error(`failed to dequeue tasks: ${stringifyError(tasks.error)}`);
```

Comment:
just a heads up, happened to me once when the API stop responding all fetch are failing fast ultimately using the whole cpu. Might be an edge case but fyi

Reply:
good point. should we introduce a little delay in case of error?